lib/src/lazy_load_scrollview.dart

```diff
@@ -1,89 +1,296 @@
-library lazy_load_scrollview;
+import 'dart:async';
 
-import 'package:flutter/widgets.dart';
+import 'package:flutter/foundation.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter/rendering.dart';
 
-enum LoadingStatus { LOADING, STABLE }
+// TODO: How does this know it's reached the end of the list?
+/// A function signature for a builder that recieves loading widgets to be
+/// inserted into a ListView when further data is being fetched.
+typedef LazyLoadBuilder = Widget Function(
+  BuildContext context,
+  bool isLoadingAfter,
+  bool isLoadingBefore,
+);
 
-/// Signature for EndOfPageListeners
-typedef void EndOfPageListenerCallback();
-
-/// A widget that wraps a [Widget] and will trigger [onEndOfPage] when it
-/// reaches the bottom of the list
 class LazyLoadScrollView extends StatefulWidget {
-  /// The [Widget] that this widget watches for changes on
-  final Widget child;
+  /// A lazy builder that will be used to create the list widget. It also
+  /// carries the context of the child and a loading widget to be inserted into
+  /// the ListViews.
+  ///
+  /// Note, the direct child of this widget should be the [ListView]. If it is
+  /// not and the [ListView] is further down the tree then it's likely that this
+  /// will Builder will fail to recognize scroll events unless [ignoreDepth] is
+  /// set to true.
```

Comment:
If this is likely and undesired behaviour, can we add assertions in the constructor to prevent this from happening?

Reply:
Hmm, I think I wrote that before the predicate. How does this sound instead:

```dart
/// The [Scrollable] should be a direct child of this builder for us to listen
/// to bubbling events. If it is not the direct child, you must further specify
/// a predicate in order to find it in the descendents or, if there are no other
/// scrollables possible, set [ignoreDepth] to true. If you'd like to avoid a
/// predicate, you may also provide an optional [ScrollController].
```

I don't think we can do an assertion either of a child of a builder because it'd need to be built first too. But that comment should clear things up.